### PR TITLE
added rotate #tuvix

### PR DESCRIPTION
**Pull request by CAChristopher**

### CODE REVIEW

File: pkg/broker/logic.go [hunk 1]
```diff
@@ -39,6 +42,33 @@ func (b *BusinessLogic) GetCatalog(c *broker.RequestContext) (*broker.CatalogRes
...
+	}
+
+	user, err := provider.RotateCredentials(instance)
+	if err != nil {
+		glog.Errorf("Unable to rotate access keys, RotateCredentials failed: %s\n", err.Error())
+		return nil, InternalServerError()
+	}
+
+	err = b.storage.UpdateCredentials(instance, user)
+	if err != nil {
```
Comment (trevorlinton) [inline pkg/broker/logic.go:65]: Are we going to somehow indicate that the older credentials are deprecated or remove the old creds?

File: pkg/broker/logic.go [hunk 2]
```diff
@@ -39,6 +42,33 @@ func (b *BusinessLogic) GetCatalog(c *broker.RequestContext) (*broker.CatalogRes
...
+		glog.Errorf("Error: Unable to record password change for instance %s and user %s with new password %s\n", instance.Name, user.AccessKeyId, user.SecretAccessKey)
+		return nil, InternalServerError()
```
Comment (trevorlinton) [inline pkg/broker/logic.go:67]: Lets not write the password to logs, just knowing the name and access key id should be enough to fix the error, no?

File: pkg/broker/logic.go [hunk 3]
```diff
@@ -282,16 +312,16 @@ func (b *BusinessLogic) Update(request *osb.UpdateInstanceRequest, c *broker.Req
 
 func (b *BusinessLogic) LastOperation(request *osb.LastOperationRequest, c *broker.RequestContext) (*broker.LastOperationResponse, error) {
 	response := broker.LastOperationResponse{}
-	
+
 	upgrading, err := b.storage.IsUpgrading(request.InstanceID)
 	if err != nil {
-		glog.Errorf("Unable to get resource (%s) status, IsUpgrading failed: %s\n", request.InstanceID, err.Error()) 
+		glog.Errorf("Unable to get resource (%s) status, IsUpgrading failed: %s\n", request.InstanceID, err.Error())
 		return nil, InternalServerError()
```
Comment (trevorlinton) [inline pkg/broker/logic.go:320]: Can you please turn off auto-formatting on your IDE? It makes pull requests noisy and we agreed we would put global formatting changes into a separate PR.

File: pkg/broker/logic.go [hunk 4]
```diff
@@ -25,6 +25,9 @@ func NewBusinessLogic(ctx context.Context, o Options) (*BusinessLogic, error) {
 		storage:    storage,
 		namePrefix: namePrefix,
 	}
+
+	bl.AddActions("rotate_credentials", "credentials", "PUT", bl.ActionRotateCredentials)
+
```
Comment (trevorlinton) [inline pkg/broker/logic.go:31]: Is there a CLI plugin for S3 so users can invoke a rotation?

**Review by trevorlinton — CHANGES_REQUESTED** [review]
If we're not going to remove the old IAM credentials we need to create a way of seeing old credentials and removing them) or we need to delete the old IAM credentials (or at least mark them as inactive if possible?).

In addition I didn't see a CLI plugin accompanying this, how are users going to initiate a rotation?

Also, when requesting a review can you add the #tuvix label/tag to this? It auto notifies everyone that a PR is ready for review.

File: pkg/broker/logic.go (hunk 1 above)
Comment (CAChristopher) [reply pkg/broker/logic.go:65]: Wanted to talk to you about this. Is the creds are removed and the app not restarted then the app is kinda boned. I was thinking a task to do it after a day or two?

Comment (trevorlinton) [reply pkg/broker/logic.go:65]: You're right in that it won't restart the app or any apps that the s3 is attached too.  

We could add an additional action to list out all the access key id's and indicate which one is the one being used, then add another option to destroy a specific access key id.  That way the end user on the CLI can first rotate it, take whatever steps necessary (like restart an app) then go back and remove the old key id's once everything is up and working correctly.

Thoughts?

Comment (CAChristopher) [reply pkg/broker/logic.go:65]: yeah, as long as we are ok with the manual step to remove.  If it won't restart the app, the app will break anyway if it invalidates, so i'm almost tempted to delete the key, but there would be no rollback. So the options are, delete the old key outright, or then add a manual step in. You can only have two keys existing at a time, so they'd have to delete the old key before rotating again.

**Review by trevorlinton — APPROVED** [review]
